modules/media.js

```diff
@@ -1,72 +1,162 @@
+  } catch (e) {
+    throw e;
+  }
+
+  listFiles.onsuccess = function() {
+    var file = this.result;
+    if (file) {
+      var name = that.sanatizeFilename(file.name);
+      that.files[name] = true;
+    } else {
+      that.updating = false;
+    }
   };
 
-  // Each storage contains a store {DeviceStorage} and a ready {boolean}
-  this.external = {
-    apps: {},
-    music: {},
-    pictures: {},
-    sdcard: {},
-    videos: {}
+  listFiles.onerror = function() {
+    throw new Error('Failed to list files on storage device ' + this.type);
   };
+};
 
+/**
+ * @description Santizes file names such that they are valid
+ *   object keys.
+ * @fname File name to sanatize.
+ * @returns {string}
+ */
+Storage.prototype.sanatizeFilename = function(fname) {
+  // TODO - This is not legit.
+  fname = fname.replace(/-/g, 'x');
+  fname = fname.replace(/\./g, 'y');
+  return fname;
+};
 
-  // Private data members
 
-  var appStores; // {array of DeviceStorage}
-  var musicStores; // {array of DeviceStorage}
-  var picturesStores; // {array of DeviceStorage}
-  var sdcardStores; // {array of DeviceStorage}
-  var videosStores; // {array of DeviceStorage}
 
 
-  // Constructor
 
-  appStores = navigator.getDeviceStorages('apps');
-  musicStores = navigator.getDeviceStorages('music');
-  picturesStores = navigator.getDeviceStorages('pictures');
-  sdcardStores = navigator.getDeviceStorages('sdcard');
-  videosStores = navigator.getDeviceStorages('video');
+/**
+ * Manages internal and external storages, or handles to storage
+ * devices, and their various data sets, including apps, music,
+ * pictures, sdcard, and videos.
+ */
+function Media() {
 
+  // Public data members
 
-  // TODO -- we need a way of tracking which of the above
-  // calls failed.
+  // Valid storage types
+  this.storageTypes = [
+    // 'apps',
+    // 'music',
+    'pictures',
+    'sdcard',
+    // 'videos'
+  ];
 
+  // Contains a storage {Storage} for each internal storage type.
+  this.internal = {};
 
-  // initialize app storage
-  this.internal.apps = this.getInternalStorage(appStores);
-  this.external.apps = this.getExternalStorage(appStores);
+  // Contains a storage {Storage} for each external storage type.
+  this.external = {};
 
-  // initialize music storage
-  this.internal.music = this.getInternalStorage(musicStores);
-  this.external.music = this.getExternalStorage(musicStores);
+  // Constructor
+  for (var i = 0; i < this.storageTypes.length; ++i) {
 
-  // initialize picture storage
-  this.internal.pictures = this.getInternalStorage(picturesStores);
-  this.external.pictures = this.getExternalStorage(picturesStores);
+    var type = this.storageTypes[i];
 
-  // initialize sdcard storage
-  this.internal.sdcard = this.getInternalStorage(sdcardStores);
-  this.external.sdcard = this.getExternalStorage(sdcardStores);
+    // Fetch all {DeviceStorage} of the specified type
+    var stores = navigator.getDeviceStorages(type);
```

Comment:
What (if anything) happens if getDeviceStorages doesn't find anything of the given type?